### PR TITLE
Resume the session

### PR DESCRIPTION
**Pull request by markccchiang**
This is for issue #172. When the connection drops (due to the crash of frontend or backend, or the network problem). The session can be resumed by the frontend sending a recovering message to the backend. The message includes recovering image files, their channels, and regions on the backend. It is tricky to persist in a session for a period of time. So far the session is deleted once the `On Disconnect` is called. The timeout setting for uWcbsocket is 15 secs. That means after 15 secs of the network dropping down, the backend session will be deleted.

### CODE REVIEW

File: Main.cc [hunk 1]
```diff
@@ -109,6 +107,20 @@ void OnDisconnect(uWS::WebSocket<uWS::SERVER>* ws, int code, char* message, size
...
+    switch ((long)user) {
+        case 3:
+            cerr << "Client emitted error on connection timeout (non-SSL)" << endl;
+            break;
+        case 5:
+            cerr << "Client emitted error on connection timeout (SSL)" << endl;
+            break;
+        default:
+            cerr << "FAILURE: " << user << " should not emit error!" << endl;
+            exit(-1);
```
Comment (veggiesaurus) [inline Main.cc]: does this mean that an error in one message from one user would crash the backend for all users?

**Comment by veggiesaurus** [comment]
@kswang1029 spotted an issue when sending a resume message while calculating a cube historgram. @markccchiang could you look into this before we merge this branch in?

The code itself looks fine, so if we can just check that issue, then we can merge this in.

**Comment by kswang1029** [comment]
It is “streaming spectral profile” causing the backend crash while resuming a session. For cube histogram, a session can be resumed but cube histogram will not be streamed anymore. Sounds like similar root cause?

File: Main.cc (hunk 1 above)
Comment (markccchiang) [reply Main.cc]: I think it does. I will remove the line`exit(-1);` to avoid such a case.

**Comment by markccchiang** [comment]
While the `OnResumeSession` called, the old streaming processes including spectral profile, cube histogram, and animation should be stopped. The old message queue should be cleared. @kswang1029 could you please test the updated branch solves the problem?

**Comment by kswang1029** [comment]
@markccchiang @veggiesaurus 
streaming region spectral profile: session is resumed without crash, but region spectral profiler is not functional anymore. 

streaming animation: session is resumed without crash, but I see a persistent indicator of streaming tiles at the top-right corner. And we cannot make new regions anymore.

streaming per-cube histogram: session is resumed without crash, but once I zoom the image, I see a persistent indicator of streaming tiles at the top-right corner. And we cannot make new regions anymore.

**Comment by kswang1029** [comment]
@markccchiang @veggiesaurus 
with d2944a1, all symptoms are gone! 👍👍👍

**Review by veggiesaurus — APPROVED** [review]

**Pull request merged**